### ISSUE TITLE
The DefaultHttpClientAdapter shouldn't "remember" the validness of a host

### ISSUE THREAD

**Issue by AegoYu** [flutterchina/dio#1175]
### New Issue Checklist

- [x] I have searched for a similar issue in the [project](https://github.com/flutterchina/dio/issues) and found none

### Issue Info

| Info                           | Value                                                 |      |
| ------------------------------ | ----------------------------------------------------- | ---- |
| Platform Name                  | e.g. flutter / iOS / android                          | flutter     |
| Platform Version               | e.g.  1.5.0 / 12.0 / 9.0                              |      |
| Dio Version                    | e.g.  2.1.0 / 1.0.17                                  |   4.0.0   |
| Android Studio / Xcode Version | e.g. Android Studio 3.3.2 / Xcode 10.2.1              |      |
| Repro rate                     | e.g. all the time (100%) / sometimes x% / only once   |  all the time    |
| Repro with our demo prj        | e.g. does it happen with our demo project?            |      |
| Demo project link              | e.g. link to a demo project that highlights the issue |      |

### Issue Description and Steps

It's a scenario usually happen in the real world:

When running dio with flutter on a mobile device, the device may be disconnected at any time due to the instability of user's wifi or mobile network. When the network is disconnected, sending a HTTP request will result as a failure to lookup host. In my case, it will cause the http adapter to "remember" the result of the "invalid host". However, the host may be actually existing. So when the user connects to the Internet again, he will retry the request, and still result as a failure due to such a host cache.

Now I have to manually reset the httpClientAdapter every time when I catch a network error without a response as a workaround to solve the issue. I wonder if there is a better way to deal with.
